rcl_action/include/rcl_action/action_server.h

```diff
@@ -0,0 +1,667 @@
+ * \param[in] action_server the handle to the action server that will send the cancel response
+ * \param[in] cancel_response a struct containing the cancel response
+ * \return `RCL_RET_OK` if the request was taken, or
+ * \return `RCL_RET_INVALID_ARGUMENT` if any arguments are invalid, or
+ * \return `RCL_RET_ACTION_SERVER_INVALID` if the action server is invalid, or
+ * \return `RCL_RET_ERROR` if an unspecified error occurs.
+ */
+RCL_PUBLIC
+RCL_WARN_UNUSED
+rcl_ret_t
+rcl_action_send_cancel_response(
+  const rcl_action_server_t * action_server,
+  const rcl_action_cancel_response_t * cancel_response);
+
+/// Get the name of the action for a rcl_action_server_t.
+/**
+ * This function returns the action server's internal topic name string.
+ * This function can fail, and therefore return `NULL`, if the:
+ *   - action server is `NULL`
+ *   - action server is invalid (never called init, called fini, or invalid)
+ *
+ * The returned string is only valid as long as the action server is valid.
+ * The value of the string may change if the topic name changes, and therefore
+ * copying the string is recommended if this is a concern.
+ *
+ * <hr>
+ * Attribute          | Adherence
+ * ------------------ | -------------
+ * Allocates Memory   | No
+ * Thread-Safe        | No
+ * Uses Atomics       | No
+ * Lock-Free          | Yes
+ *
+ * \param[in] action_server the pointer to the action server
+ * \return name string if successful, otherwise `NULL`
+ */
+RCL_PUBLIC
+RCL_WARN_UNUSED
+const char *
+rcl_action_server_get_action_name(const rcl_action_server_t * action_server);
+
+/// Return the rcl_action_server_options_t for a rcl_action_server_t.
+/**
+ * This function returns the action server's internal options struct.
+ * This function can fail, and therefore return `NULL`, if the:
+ *   - action server is `NULL`
+ *   - action server is invalid (never called init, called fini, or invalid)
+ *
+ * The returned struct is only valid as long as the action server is valid.
+ * The values in the struct may change if the action server's options change,
+ * and therefore copying the struct is recommended if this is a concern.
+ *
+ * <hr>
+ * Attribute          | Adherence
+ * ------------------ | -------------
+ * Allocates Memory   | No
+ * Thread-Safe        | No
+ * Uses Atomics       | No
+ * Lock-Free          | Yes
+ *
+ * \param[in] action_server pointer to the action server
+ * \return options struct if successful, otherwise `NULL`
+ */
+RCL_PUBLIC
+RCL_WARN_UNUSED
+const rcl_action_server_options_t *
+rcl_action_server_get_options(const rcl_action_server_t * action_server);
+
+/// Return the goal handles for all active or terminated goals.
+/**
+ * A pointer to the internally held array of goal handle structs is returned
+ * along with the number of items in the array.
+ * Goals that have terminated, successfully responded to a client with a
+ * result, and have expired (timed out) are not present in the array.
```

Comment:
Who controls the timeout. Is there a function to check if the goals are expired? Is there a timer part of the action server?

Reply:
I'll add a function for checking timeouts. I imagine it would be better to implement this logic in the action server so it isn't duplicated in the client libraries.

Reply:
+1 to implementing it in the action server